text/0131-pass-native-sdk-spans-to-hybrid.md

```diff
@@ -0,0 +1,174 @@
+Furthermore, there could be native-only operations that our hybrid SDKs currently don't capture. Therefore,
+we want to pass native spans up to the native layers so users know what's happening on the native layer.
+
+Furthermore, the hybrid SDKs can then establish a parent-child relationship between the spans. For
+example, if there is only one HTTP span on the hybrid layer and one HTTP span on the native layer
+that occurred within the lifetime of the hybrid HTTP span, the hybrid SDK can set the native span as
+a child of the hybrid span. Of course, the hybrid SDK must also compare the URL in that case or
+apply other logic to avoid wrong parent-child relationships.
+
+# Background
+
+The mobile Starfish team wants to have more data for hybrid SDKs to improve the value of our
+performance offering. @shruthilayaj brought up this topic several times during mobile Starfish team
+syncs. After (@krystofwoldrich and @philipphofmann) hacked together a POC for HTTP spans from Cocoa
+to RN, we decided this would be a valuable feature. You can visit the code on
+[Cocoa](https://github.com/getsentry/sentry-cocoa/tree/poc/network-tracker-finished-spans) and
+[RN](https://github.com/getsentry/sentry-react-native/tree/kw-cocoa-spans-to-rn-poc) and the internal
+[Notion doc](https://www.notion.so/sentry/RN-Span-ID-to-Cocoa-POC-085194074d39415684eb3d0e37b70c99).
+
+The solution has to support the following requirements:
+
+1. It should be compatible with both transactions and single-span ingestion. It must be compatible
+with single-span ingestion. Not being compatible with transactions might be acceptable, as we have
+plans to remove them.
+2. It must work with transactions bound to the scope and active spans. We don't have to worry about
+multiple transactions or spans running in parallel, as on mobile, the SDKs can never tell to which
+transaction bound to the scope or active span an auto-instrumented span belongs. Therefore, they
+attach it to the transaction bound to the scope or the active span.
+3. Native auto-instrumented spans must go through the hybrid layer because of span filtering and sampling,
+meaning native SDKs must not send spans themselves.
+4. The native SDKs must support attaching screen frame data to the spans.
+
+# Options Considered
+
+## Option 1: Self Clearing Cache With Hooks <a name="option-1"></a>
+
+The native SDKs implement a self-clearing cache for their auto-instrumented spans with hooks to
+start and stop the cache. Whenever the native SDKs add a span to their cache, they discard all
+spans older than a configurable threshold. To make cache clearing efficient, the SDKs keep the spans
+in a data structure ordered by span timestamp and use a binary search to find the last index of the
+range of spans to delete, which makes inserting a span `O(log n)`. Furthermore, the native SDKs
+offer a callback when all ongoing spans are finished to support `waitForChildren`.
```

Comment:
> to support `waitForChildren`
The `waitForChildren` is from the Hybrid SDKs, is that right?

Reply:
Yes